### PR TITLE
Add Modelica code to ThermalFluid Benchmark

### PR DESCRIPTION
**Pull request by ChrisRackauckas**

### CODE REVIEW

File: .buildkite/build_benchmark.sh [hunk 1]
```diff
@@ -13,6 +14,23 @@ if [[ "${JULIAHUBREGISTRY_BENCHMARK_TARGETS[*]}" =~ "${1}" ]]; then
...
 
+if [[ "${OPENMODELICA_BENCHMARK_TARGETS[*]}" =~ "${1}" ]]; then
+	echo "--- :toolbox: Installing OpenModelica"
+	adduser --force-badname --system --home /nonexistent  \
+            --no-create-home --quiet _apt
+	apt-get update
+	apt-get install --yes ca-certificates curl gnupg
+	curl -fsSL http://build.openmodelica.org/apt/openmodelica.asc | gpg --dearmor -o /usr/share/keyrings/openmodelica-keyring.gpg
+
+	echo "deb [arch=arm64 signed-by=/usr/share/keyrings/openmodelica-keyring.gpg] https://build.openmodelica.org/apt \
```
Comment (staticfloat) [inline .buildkite/build_benchmark.sh:25]:
```suggestion
	echo "deb [arch=amd64 signed-by=/usr/share/keyrings/openmodelica-keyring.gpg] https://build.openmodelica.org/apt \
```

We were downloading the package listings for the wrong architecture.  :P

File: .buildkite/build_benchmark.sh [hunk 2]
```diff
@@ -13,6 +14,23 @@ if [[ "${JULIAHUBREGISTRY_BENCHMARK_TARGETS[*]}" =~ "${1}" ]]; then
 	julia -e 'using Pkg; Pkg.Registry.add(); Pkg.Registry.status()'
 fi
 
+if [[ "${OPENMODELICA_BENCHMARK_TARGETS[*]}" =~ "${1}" ]]; then
+	echo "--- :toolbox: Installing OpenModelica"
+	adduser --force-badname --system --home /nonexistent  \
+            --no-create-home --quiet _apt
```
Comment (staticfloat) [inline .buildkite/build_benchmark.sh:20]:
```suggestion
```

This should not be needed.

File: .buildkite/build_benchmark.sh (hunk 1 above)
Comment (thazhemadam) [reply .buildkite/build_benchmark.sh]: Good catch!

**Comment by ChrisRackauckas** [comment]
We're going to have to `@elapsed` OpenModelica to keep it honest. One of the cases literally runs for over 10 minutes and then it reports out it took a second. Clearly it doesn't measure compile times...

File: benchmarks/ModelingToolkit/ThermalFluid.jmd [hunk 3]
```diff
@@ -1,16 +1,22 @@
...
-This is a 1D advection-diffusion-source PDE that uses a second order upwind scheme. 
+This is a 1D advection-diffusion-source PDE that uses a second order upwind scheme.
+
 ```julia
 using Pkg
 # Rev fixes precompilation https://github.com/hzgzh/XSteam.jl/pull/2
 Pkg.add(Pkg.PackageSpec(;name="XSteam", rev="f2a1c589054cfd6bba307985a3a534b6f5a1863b"))
 
 using ModelingToolkit, JuliaSimCompiler, Symbolics, XSteam, Polynomials, BenchmarkTools, CairoMakie
+using OpenModelica, OrdinaryDiffEq
```
Comment (ChrisRackauckas) [inline benchmarks/ModelingToolkit/ThermalFluid.jmd:14]:
```suggestion
using OMJulia, OrdinaryDiffEq
```

**Comment by ChrisRackauckas** [comment]
@chriselrod do you know what

```
<html>
<body>
<!--StartFragment-->


warning: Linking two modules of different target triples: 'start' is 'x86_64-linux-gnu' whereas 'scimlde' is 'x86_64-unknown-linux-gnu'
  |  
  | warning: Linking two modules of different target triples: '' is 'x86_64-linux-gnu' whereas 'scimlde' is 'x86_64-unknown-linux-gnu'
 

<br class="Apple-interchange-newline"><!--EndFragment-->
</body>
</html>
```

is all about? It seems to be stalling on that before it even gets to the benchmarks.

**Comment by chriselrod** [comment]
https://github.com/JuliaComputing/JuliaSimCompiler.jl/issues/273
I assume it is doing that while running the JuliaSimCompiler benchmarks. I think it probably takes very little time compared to the of the compilation.

**Comment by ChrisRackauckas** [comment]
The model is showing the result diverging: https://buildkite.com/julialang/scimlbenchmarks-dot-jl/builds/2412#018fe880-ac22-45fe-b7c0-f09cc1fce9f9/736-1408. You validated the model gives the right output?

File: benchmarks/ModelingToolkit/ThermalFluid.jmd [hunk 4]
```diff
@@ -278,6 +282,33 @@ function build_run_problem(fsys, N; target=JuliaSimCompiler.JuliaTarget())
...
+            prob = ODEProblem(sys, target, [], tspan, sparse=true)
+        end
+        prob.u0 .= 12.0
+        solve(prob, solver, reltol=1e-6, abstol=1e-6, saveat=100);
+    end
+end
+```
+
+```julia
+N_x = [5, 10, 20, 40, 60, 80, 160, 320, 480, 640, 800, 960, 1280];
```
Comment (ChrisRackauckas) [inline benchmarks/ModelingToolkit/ThermalFluid.jmd:302]:
```suggestion
N_x = [5, 10, 20, 40, 60, 80, 160];
```

File: benchmarks/ModelingToolkit/ThermalFluid.jmd [hunk 5]
```diff
@@ -288,41 +320,112 @@ const LLVMBackend = JuliaSimCompiler.llvm.LLVMTarget();
...
+  total_times[i, 4] = test_speed(IRSystem, N_x_i, target=LLVMBackend)
+
+  @show N_x_i, ss_times[i, :], times[i, :], total_times[i, :]
+end
+```
+
+## Time OpenModelica
+
+```julia
+using OMJulia, CSV, DataFrames, Plots
```
Comment (ChrisRackauckas) [inline benchmarks/ModelingToolkit/ThermalFluid.jmd:346]:
```suggestion
using OMJulia, CSV, DataFrames
```

**Comment by ChrisRackauckas** [comment]
```
Error: OMCError 
Command `omc --interactive=zmq -z=julia.KutJRULdDd` failed
Error: You are trying to run OpenModelica as a server using the root user.
This is a very bad idea:
* The socket interface does not authenticate the user.
* OpenModelica allows execution of arbitrary commands.

Execution failed!
```

@thazhemadam can you help with that?

File: benchmarks/ModelingToolkit/ThermalFluid.jmd (hunk 4 above)
Comment (chriselrod) [reply benchmarks/ModelingToolkit/ThermalFluid.jmd]: On my desktop, MTK's `structural_simplify` took more than 10 minutes for the largest size.

Comment (ChrisRackauckas) [reply benchmarks/ModelingToolkit/ThermalFluid.jmd]: Should we do a cutoff on pure MTK?

Comment (chriselrod) [reply benchmarks/ModelingToolkit/ThermalFluid.jmd:302]: Yeah, I think that's a good idea, depending on how well OpenModelica can handle larger sizes/finer discritizations.

It's nice to go further for the faster methods to make claims like "with fixed time, here is how many discretized components you can use".

Comment (ChrisRackauckas) [reply benchmarks/ModelingToolkit/ThermalFluid.jmd:302]: okay, I got this all working except the OpenModelica part now.

File: benchmarks/ModelingToolkit/ThermalFluid.jmd [hunk 6]
```diff
@@ -288,41 +319,119 @@ const LLVMBackend = JuliaSimCompiler.llvm.LLVMTarget();
...
+        ModelicaSystem(mod, modelicafile, "DhnControl.Test.test_preinsulated_470_$N")
+        sendExpression(mod, "simulate(DhnControl.Test.test_preinsulated_470_$N)")
+    end
+    #runtime = res["timeTotal"]
+    @assert res["messages"][1:11] == "LOG_SUCCESS"
+    total_times[i, 5] = totaltime
+end
+
+OMJulia.quit(mod)
+=#
```
Comment (ChrisRackauckas) [inline benchmarks/ModelingToolkit/ThermalFluid.jmd:370]:
```suggestion
using OMJulia, CSV, DataFrames
mod = OMJulia.OMCSession();
OMJulia.sendExpression(mod, "getVersion()")
modelicafile = "../../benchmarks/ModelingToolkit/DhnControl.mo"
resultfile = "modelica_res.csv"

@show "Start OpenModelica Timings"

for N in N_x
    @show N
    totaltime = @elapsed res = begin
        ModelicaSystem(mod, modelicafile, "DhnControl.Test.test_preinsulated_470_$N")
        sendExpression(mod, "simulate(DhnControl.Test.test_preinsulated_470_$N)")
    end
    #runtime = res["timeTotal"]
    @assert res["messages"][1:11] == "LOG_SUCCESS"
    total_times[i, 5] = totaltime
end

OMJulia.quit(mod)
```

File: benchmarks/ModelingToolkit/DhnControl.mo [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+    model test_preinsulated_470_960
+        extends DhnControl.Test.test_preinsulated_470(preinsulatedPipe(N=960));
+    end test_preinsulated_470_960;
+
+    model test_preinsulated_470_1280
+        extends DhnControl.Test.test_preinsulated_470(preinsulatedPipe(N=1280));
+    end test_preinsulated_470_1280;
+
+  end Test;
+end DhnControl;
```
Comment (avinashresearch1) [inline benchmarks/ModelingToolkit/DhnControl.mo:295]:
```
 package TimingFunctions

    function tic
      "Function to record the internal time in [s]"
      output Real tic;
    algorithm
      (ms_tic, sec_tic, min_tic, hour_tic, day_tic, mon_tic, year_tic) := Modelica.Utilities.System.getTime();
      tic := (ms_tic*0.001) + sec_tic + (min_tic*60) + (hour_tic*3600) + (day_tic*86400);
    end tic;

  end TimingFunctions;
```
```suggestion
end DhnControl;
```

Comment (avinashresearch1) [inline benchmarks/ModelingToolkit/DhnControl.mo:295]:
```suggestion
 package TimingFunctions
    function tic
      "Function to record the internal time in [s]"
      output Real tic;
    algorithm
      (ms_tic, sec_tic, min_tic, hour_tic, day_tic, mon_tic, year_tic) := Modelica.Utilities.System.getTime();
      tic := (ms_tic*0.001) + sec_tic + (min_tic*60) + (hour_tic*3600) + (day_tic*86400);
    end tic;

  end TimingFunctions;
end DhnControl;
```

Comment (avinashresearch1) [reply benchmarks/ModelingToolkit/DhnControl.mo:295]: This is the `tic` used in the .mos file for consistency.

File: benchmarks/ModelingToolkit/ThermalFluid.jmd [hunk 8]
```diff
@@ -288,41 +319,117 @@ const LLVMBackend = JuliaSimCompiler.llvm.LLVMTarget();
...
+
+```julia
+f = Figure(size = (800, 1200));
+let ax = Axis(f[1, 1]; yscale = log10, xscale = log10, title="Structural Simplify Time")
+  names = ["MTK", "JSIR"]
+  _lines = map(enumerate(names)) do (j, label)
     ts = @view(ss_times[:, j])
-    lines!(N_states, ts; label)
+    lines!(N_states, ts)
   end
```
Comment (chriselrod) [inline benchmarks/ModelingToolkit/ThermalFluid.jmd:404]:
```suggestion
  _lines = map(eachcol(ss_times)) do ts
    lines!(N_states, ts)
  end
```

File: benchmarks/ModelingToolkit/ThermalFluid.jmd [hunk 9]
```diff
@@ -288,41 +319,117 @@ const LLVMBackend = JuliaSimCompiler.llvm.LLVMTarget();
...
   title = timecat * " Time"
-  ax = Axis(f[i+1, 1]; title)
-  for (j, label) in enumerate(("MTK", "JSIR - Julia", "JSIR - C", "JSIR - LLVM"))
+  ax = Axis(f[i+1, 1]; yscale = log10, xscale = log10, title)
+  names = ["MTK", "JSIR - Julia", "JSIR - C", "JSIR - LLVM"]
+  _lines = map(enumerate(names)) do (j, label)
     ts = getindex.(@view(times[:, j]), i)
-    lines!(N_states, ts; label)
+    lines!(N_states, ts)
   end
```
Comment (chriselrod) [inline benchmarks/ModelingToolkit/ThermalFluid.jmd:414]:
```suggestion
  _lines = map(eachcol(times)) do ts
    lines!(N_states, getindex.(ts, i))
  end
```

**Pull request merged**